src/lib.rs

```diff
@@ -0,0 +1,414 @@
+                if c == 0 {
+                    return None;
+                }
+                if c == cnt.compare_and_swap(c, c - 1, Ordering::AcqRel) {
+                    break;
+                }
+            }
+        }
+        Some(self.task.clone())
+    }
+}
+
+fn partition(s: &str, pattern: char) -> (&str, Option<&str>) {
+    let mut splits = s.splitn(2, pattern);
+    (splits.next().unwrap(), splits.next())
+}
+
+impl FromStr for Action {
+    type Err = String;
+
+    /// Parse an action.
+    ///
+    /// `s` should be in the format `[p%][cnt*]task[(args)]`, `p%` is the frequency,
+    /// `cnt` is the max times the action can be triggered.
+    fn from_str(s: &str) -> Result<Action, String> {
+        let mut remain = s.trim();
+        let mut args = None;
+        // in case there is '%' in args, we need to parse it first.
+        let (first, second) = partition(remain, '(');
+        if let Some(second) = second {
+            remain = first;
+            if !second.ends_with(")") {
+                return Err("parentheses not match".to_owned());
+            }
+            args = Some(&second[..second.len() - 1]);
+        }
+
+        let mut frequency = 1f32;
+        let (first, second) = partition(remain, '%');
+        if let Some(second) = second {
+            remain = second;
+            match first.parse::<f32>() {
+                Err(e) => return Err(format!("failed to parse frequency: {}", e)),
+                Ok(freq) => frequency = freq / 100.0,
+            }
+        }
+
+        let mut max_cnt = None;
+        let (first, second) = partition(remain, '*');
+        if let Some(second) = second {
+            remain = second;
+            match first.parse() {
+                Err(e) => return Err(format!("failed to parse count: {}", e)),
+                Ok(cnt) => max_cnt = Some(cnt),
+            }
+        }
+
+        let parse_timeout = || match args {
+            None => return Err("sleep require timeout".to_owned()),
+            Some(timeout_str) => match timeout_str.parse() {
+                Err(e) => return Err(format!("failed to parse timeout: {}", e)),
+                Ok(timeout) => Ok(timeout),
+            },
+        };
+
+        let task = match remain {
+            "off" => Task::Off,
+            "return" => Task::Return(args.map(str::to_owned)),
+            "sleep" => Task::Sleep(try!(parse_timeout())),
+            "panic" => Task::Panic(args.map(str::to_owned)),
+            "print" => Task::Print(args.map(str::to_owned)),
+            "pause" => Task::Pause,
+            "yield" => Task::Yield,
+            "delay" => Task::Delay(try!(parse_timeout())),
+            _ => return Err(format!("unrecognized command {:?}", remain)),
+        };
+
+        Ok(Action::new(task, frequency, max_cnt))
+    }
+}
+
+// A better name?
+pub type ReturnValue = Option<String>;
+
+pub struct FailPoint {
+    name: &'static str,
+    pause: Mutex<bool>,
+    pause_notifier: Condvar,
```

Comment:
may be easier to use thread pack and unpack directly.

Reply:
Then I need to use a `mutex + vector` to save the thread handles. I think it's more simple to use `Mutex<bool> + Condvar`.